File: crates/devplat/src/adapter_execution_evidence.rs
//! Digest-addressed caller evidence for concrete adapter execution.
//!
//! The adapter registry declares routes; the Python and native adapter layers perform the
//! actual parsing/conformance work. This module is the narrow handoff between those planes. It
//! retains what a caller observed about one selected adapter, including explicit refusal,
//! conformance, output identity, and semantic-loss posture, without making the MCP core execute
//! Python, import packages, open locators, or certify scientific meaning.

use bioprism_ids::ContentHash;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::BTreeSet;

pub const ADAPTER_EXECUTION_EVIDENCE_SCHEMA: &str =
    "bioprism-devplat-adapter-execution-evidence/0.1";
pub const ADAPTER_EXECUTION_EVIDENCE_WORKFLOW: &str = "adapter_execution_evidence";
pub const MAX_ADAPTER_EXECUTION_EVIDENCE_DOMAINS: usize = 64;
pub const MAX_ADAPTER_EXECUTION_EVIDENCE_LOSSES: usize = 128;
pub const MAX_ADAPTER_EXECUTION_EVIDENCE_PARENTS: usize = 128;
pub const MAX_ADAPTER_EXECUTION_EVIDENCE_TEXT_BYTES: usize = 512;
pub const MAX_ADAPTER_EXECUTION_EVIDENCE_BYTES: u64 = 68_719_476_736;
pub const MAX_ADAPTER_EXECUTION_EVIDENCE_ITEMS: u64 = 2_000_000;

const EXECUTION_STATUSES: &[&str] = &[
    "planned",
    "started",
    "succeeded",
    "partial",
    "refused",
    "failed",
    "unknown",
];
const CONFORMANCE_STATUSES: &[&str] = &["verified", "partial", "refused", "not_run", "unknown"];
const SEMANTIC_LOSS_STATUSES: &[&str] = &["lossless", "lossy", "unknown", "not_applicable"];
const LOSS_SEVERITIES: &[&str] = &["info", "warning", "blocking"];

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct AdapterExecutionLoss {
    pub kind: String,
    pub severity: String,
    pub detail: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct AdapterExecutionEvidenceRequest {
    pub group_id: String,
    pub domains: Vec<String>,
    pub subject_id: String,
    pub adapter_id: String,
    pub adapter_version: String,
    pub source_id: String,
    pub input_digest: String,
    #[serde(default)]
    pub output_digest: Option<String>,
    pub execution_status: String,
    pub conformance_status: String,
    pub semantic_loss_status: String,
    #[serde(default)]
    pub losses: Vec<AdapterExecutionLoss>,
    #[serde(default)]
    pub item_count: Option<u64>,
    #[serde(default)]
    pub byte_length: Option<u64>,
    #[serde(default)]
    pub error_code: Option<String>,
    #[serde(default)]
    pub parent_digests: Vec<String>,
    #[serde(default)]
    pub attempt_id: Option<String>,
}

fn text(field: &str, value: &str, maximum: usize) -> Result<String, String> {
    if value.trim().is_empty() {
        return Err(format!("{field} must be non-empty"));
    }
    if value.len() > maximum {
        return Err(format!("{field} exceeds the {maximum}-byte bound"));
    }
    if value.chars().any(|character| character.is_control()) {
        return Err(format!("{field} must not contain control characters"));
    }
    Ok(value.to_string())
}

fn digest(field: &str, value: &str) -> Result<String, String> {
    ContentHash::parse(value.to_string())
        .map(|digest| digest.to_string())
        .map_err(|_| format!("{field} must be a lowercase 64-character SHA-256 digest"))
}

fn optional_digest(field: &str, value: &Option<String>) -> Result<Option<String>, String> {
    value
        .as_deref()
        .map(|value| digest(field, value))
        .transpose()
}

fn validate_losses(losses: &[AdapterExecutionLoss]) -> Result<(), String> {
    if losses.len() > MAX_ADAPTER_EXECUTION_EVIDENCE_LOSSES {
        return Err(format!(
            "losses must contain at most {} entries",
            MAX_ADAPTER_EXECUTION_EVIDENCE_LOSSES
        ));
    }
    let mut identities = BTreeSet::new();
    for loss in losses {
        let kind = text("loss.kind", &loss.kind, 128)?;
        let severity = text("loss.severity", &loss.severity, 32)?;
        if !LOSS_SEVERITIES.contains(&severity.as_str()) {
            return Err(format!(
                "loss.severity must be one of {}",
                LOSS_SEVERITIES.join(", ")
            ));
        }
        let detail = text("loss.detail", &loss.detail, 512)?;
        let source_path = loss
            .source_path
            .as_deref()
            .map(|value| text("loss.source_path", value, 512))
            .transpose()?;
        let target_path = loss
            .target_path
            .as_deref()
            .map(|value| text("loss.target_path", value, 512))
            .transpose()?;
        let identity = json!({
            "kind": kind,
            "severity": severity,
            "detail": detail,
            "source_path": source_path,
            "target_path": target_path,
        });
        if !identities.insert(identity.to_string()) {
            return Err("loss entries must be unique".into());
        }
    }
    Ok(())
}

fn validate_request(request: &AdapterExecutionEvidenceRequest) -> Result<(), String> {
    text(
        "group_id",
        &request.group_id,
        MAX_ADAPTER_EXECUTION_EVIDENCE_TEXT_BYTES,
    )?;
    if request.domains.is_empty() || request.domains.len() > MAX_ADAPTER_EXECUTION_EVIDENCE_DOMAINS
    {
        return Err(format!(
            "domains must contain 1..={} entries",
            MAX_ADAPTER_EXECUTION_EVIDENCE_DOMAINS
        ));
    }
    for domain in &request.domains {
        text("domain", domain, MAX_ADAPTER_EXECUTION_EVIDENCE_TEXT_BYTES)?;
    }
    text(
        "subject_id",
        &request.subject_id,
        MAX_ADAPTER_EXECUTION_EVIDENCE_TEXT_BYTES,
    )?;
    text("adapter_id", &request.adapter_id, 256)?;
    text("adapter_version", &request.adapter_version, 128)?;
    text(
        "source_id",
        &request.source_id,
        MAX_ADAPTER_EXECUTION_EVIDENCE_TEXT_BYTES,
    )?;
    digest("input_digest", &request.input_digest)?;
    let output_digest = optional_digest("output_digest", &request.output_digest)?;
    text("execution_status", &request.execution_status, 32)?;
    if !EXECUTION_STATUSES.contains(&request.execution_status.as_str()) {
        return Err(format!(
            "execution_status must be one of {}",
            EXECUTION_STATUSES.join(", ")
        ));
    }
    text("conformance_status", &request.conformance_status, 32)?;
    if !CONFORMANCE_STATUSES.contains(&request.conformance_status.as_str()) {
        return Err(format!(
            "conformance_status must be one of {}",
            CONFORMANCE_STATUSES.join(", ")
        ));
    }
    text("semantic_loss_status", &request.semantic_loss_status, 32)?;
    if !SEMANTIC_LOSS_STATUSES.contains(&request.semantic_loss_status.as_str()) {
        return Err(format!(
            "semantic_loss_status must be one of {}",
            SEMANTIC_LOSS_STATUSES.join(", ")
        ));
    }
    validate_losses(&request.losses)?;
    if request.semantic_loss_status == "lossless" && !request.losses.is_empty() {
        return Err("lossless evidence cannot contain loss entries".into());
    }
    if request.semantic_loss_status == "lossy" && request.losses.is_empty() {
        return Err("lossy evidence must contain at least one loss entry".into());
    }
    if request.semantic_loss_status == "not_applicable" && !request.losses.is_empty() {
        return Err("not_applicable semantic loss cannot contain loss entries".into());
    }
    if request.execution_status == "succeeded" && output_digest.is_none() {
        return Err("succeeded execution requires output_digest".into());
    }
    if matches!(request.execution_status.as_str(), "refused" | "failed")
        && request.error_code.is_none()
    {
        return Err("refused or failed execution requires error_code".into());
    }
    if let Some(error_code) = &request.error_code {
        text("error_code", error_code, 128)?;
    }
    if request
        .item_count
        .is_some_and(|value| value > MAX_ADAPTER_EXECUTION_EVIDENCE_ITEMS)
    {
        return Err(format!(
            "item_count exceeds {}",
            MAX_ADAPTER_EXECUTION_EVIDENCE_ITEMS
        ));
    }
    if request
        .byte_length
        .is_some_and(|value| value > MAX_ADAPTER_EXECUTION_EVIDENCE_BYTES)
    {
        return Err(format!(
            "byte_length exceeds {}",
            MAX_ADAPTER_EXECUTION_EVIDENCE_BYTES
        ));
    }
    if request.parent_digests.len() > MAX_ADAPTER_EXECUTION_EVIDENCE_PARENTS {
        return Err(format!(
            "parent_digests must contain at most {} entries",
            MAX_ADAPTER_EXECUTION_EVIDENCE_PARENTS
        ));
    }
    for parent in &request.parent_digests {
        digest("parent_digest", parent)?;
    }
    if let Some(attempt_id) = &request.attempt_id {
        text("attempt_id", attempt_id, 128)?;
    }
    Ok(())
}

/// Canonicalize caller-supplied adapter observations into a digest-bound evidence envelope.
pub fn record_adapter_execution_evidence(
    request: AdapterExecutionEvidenceRequest,
) -> Result<Value, String> {
    validate_request(&request)?;
    let output_digest = optional_digest("output_digest", &request.output_digest)?;
    let evidence_without_digest = json!({
        "schema": ADAPTER_EXECUTION_EVIDENCE_SCHEMA,
        "workflow": ADAPTER_EXECUTION_EVIDENCE_WORKFLOW,
        "group_id": request.group_id,
        "domains": request.domains,
        "subject_id": request.subject_id,
        "adapter_id": request.adapter_id,
        "adapter_version": request.adapter_version,
        "source_id": request.source_id,
        "input_digest": request.input_digest,
        "output_digest": output_digest,
        "execution_status": request.execution_status,
        "conformance_status": request.conformance_status,
        "semantic_loss_status": request.semantic_loss_status,
        "losses": request.losses,
        "item_count": request.item_count,
        "byte_length": request.byte_length,
        "error_code": request.error_code,
        "parent_digests": request.parent_digests,
        "attempt_id": request.attempt_id,
        "attestation_posture": "caller_asserted",
    });
    let evidence_digest = ContentHash::of_value(&evidence_without_digest)
        .map_err(|error| format!("cannot digest adapter execution evidence: {error}"))?
        .to_string();
    let mut evidence = evidence_without_digest;
    evidence["evidence_digest"] = json!(evidence_digest);
    let result = json!({
        "ok": true,
        "schema": ADAPTER_EXECUTION_EVIDENCE_SCHEMA,
        "workflow": ADAPTER_EXECUTION_EVIDENCE_WORKFLOW,
        "evidence": evidence,
        "evidence_digest": evidence_digest,
        "attestation_posture": "caller_asserted",
        "execution": "not_started",
        "readiness_claimed": false,
        "guarantees": [
            "the selected adapter, source, input identity, and caller outcome are retained together",
            "conformance and semantic-loss status remain independent and explicit",
            "the evidence digest covers every supplied metadata field except its own digest"
        ],
        "does_not_claim": [
            "the MCP core executed an adapter or imported an optional dependency",
            "caller_asserted execution_status or conformance_status proves adapter correctness",
            "lossless or verified posture proves scientific, clinical, provenance, regulatory, release, or readiness validity"
        ]
    });
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request() -> AdapterExecutionEvidenceRequest {
        AdapterExecutionEvidenceRequest {
            group_id: "biological_domains".into(),
            domains: vec!["genomics".into()],
            subject_id: "subject-1".into(),
            adapter_id: "bioprism.python.vcf_text".into(),
            adapter_version: "0.1.0".into(),
            source_id: "vcf-1".into(),
            input_digest: "a".repeat(64),
            output_digest: Some("b".repeat(64)),
            execution_status: "succeeded".into(),
            conformance_status: "verified".into(),
            semantic_loss_status: "lossless".into(),
            losses: vec![],
            item_count: Some(4),
            byte_length: Some(128),
            error_code: None,
            parent_digests: vec!["c".repeat(64)],
            attempt_id: Some("attempt-1".into()),
        }
    }

    #[test]
    fn evidence_is_digest_bound_and_non_executing() {
        let result = record_adapter_execution_evidence(request()).unwrap();
        assert!(result["ok"].as_bool().unwrap());
        assert_eq!(result["execution"], "not_started");
        assert!(!result["readiness_claimed"].as_bool().unwrap());
        assert_eq!(
            result["evidence"]["evidence_digest"],
            result["evidence_digest"]
        );
    }

    #[test]
    fn loss_and_refusal_states_require_explicit_evidence() {
        let mut request = request();
        request.semantic_loss_status = "lossy".into();
        assert!(record_adapter_execution_evidence(request.clone()).is_err());
        request.semantic_loss_status = "unknown".into();
        request.execution_status = "refused".into();
        assert!(record_adapter_execution_evidence(request).is_err());
    }
}

File: python/prism_sdk/adapter_execution_evidence.py
"""Typed caller evidence for native and Python-delegated adapter execution."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from .artifacts import _digest, _mapping, _text
from .capability import _route_text, _tool_payload
from .errors import ArgumentError

ADAPTER_EXECUTION_EVIDENCE_SCHEMA = "bioprism-devplat-adapter-execution-evidence/0.1"
ADAPTER_EXECUTION_EVIDENCE_WORKFLOW = "adapter_execution_evidence"
MAX_ADAPTER_EXECUTION_EVIDENCE_LOSSES = 128
MAX_ADAPTER_EXECUTION_EVIDENCE_PARENTS = 128
MAX_ADAPTER_EXECUTION_EVIDENCE_ITEMS = 2_000_000
MAX_ADAPTER_EXECUTION_EVIDENCE_BYTES = 68_719_476_736

EXECUTION_STATUSES = frozenset({"planned", "started", "succeeded", "partial", "refused", "failed", "unknown"})
CONFORMANCE_STATUSES = frozenset({"verified", "partial", "refused", "not_run", "unknown"})
SEMANTIC_LOSS_STATUSES = frozenset({"lossless", "lossy", "unknown", "not_applicable"})
LOSS_SEVERITIES = frozenset({"info", "warning", "blocking"})


def _bounded_optional_text(name: str, value: Any, maximum: int = 512) -> str | None:
    if value is None:
        return None
    value = _text(name, value)
    if len(value.encode("utf-8")) > maximum:
        raise ArgumentError(f"{name} exceeds the {maximum}-byte bound")
    if any(ord(character) < 0x20 for character in value):
        raise ArgumentError(f"{name} must not contain control characters")
    return value


@dataclass(frozen=True)
class AdapterExecutionLoss:
    kind: str
    severity: str
    detail: str
    source_path: str | None = None
    target_path: str | None = None

    def __post_init__(self) -> None:
        _bounded_optional_text("loss kind", self.kind, 128)
        if self.severity not in LOSS_SEVERITIES:
            raise ArgumentError("loss severity is invalid")
        _bounded_optional_text("loss detail", self.detail, 512)
        _bounded_optional_text("loss source_path", self.source_path, 512)
        _bounded_optional_text("loss target_path", self.target_path, 512)

    @classmethod
    def from_wire(cls, value: Mapping[str, Any]) -> "AdapterExecutionLoss":
        raw = _mapping("adapter execution loss", value)
        allowed = {"kind", "severity", "detail", "source_path", "target_path"}
        unknown = sorted(set(raw) - allowed)
        if unknown:
            raise ArgumentError(f"adapter execution loss contains unsupported fields: {', '.join(unknown)}")
        return cls(
            kind=_bounded_optional_text("loss kind", raw.get("kind"), 128) or "",
            severity=_route_text("loss severity", raw.get("severity")),
            detail=_bounded_optional_text("loss detail", raw.get("detail"), 512) or "",
            source_path=_bounded_optional_text("loss source_path", raw.get("source_path"), 512),
            target_path=_bounded_optional_text("loss target_path", raw.get("target_path"), 512),
        )

    def to_wire(self) -> dict[str, Any]:
        result: dict[str, Any] = {"kind": self.kind, "severity": self.severity, "detail": self.detail}
        if self.source_path is not None:
            result["source_path"] = self.source_path
        if self.target_path is not None:
            result["target_path"] = self.target_path
        return result


@dataclass(frozen=True)
class AdapterExecutionEvidenceRequest:
    group_id: str
    domains: tuple[str, ...]
    subject_id: str
    adapter_id: str
    adapter_version: str
    source_id: str
    input_digest: str
    execution_status: str
    conformance_status: str
    semantic_loss_status: str
    output_digest: str | None = None
    losses: tuple[AdapterExecutionLoss, ...] = ()
    item_count: int | None = None
    byte_length: int | None = None
    error_code: str | None = None
    parent_digests: tuple[str, ...] = ()
    attempt_id: str | None = None

    def __post_init__(self) -> None:
        for name, value in (
            ("group_id", self.group_id),
            ("subject_id", self.subject_id),
            ("adapter_id", self.adapter_id),
            ("adapter_version", self.adapter_version),
            ("source_id", self.source_id),
        ):
            _bounded_optional_text(name, value, 512 if name not in {"adapter_id", "adapter_version"} else 256)
        if not self.domains or len(self.domains) > 64:
            raise ArgumentError("domains must contain 1..=64 entries")
        for domain in self.domains:
            _bounded_optional_text("domain", domain, 512)
        _digest("adapter execution input_digest", self.input_digest)
        if self.output_digest is not None:
            _digest("adapter execution output_digest", self.output_digest)
        if self.execution_status not in EXECUTION_STATUSES:
            raise ArgumentError("execution_status is invalid")
        if self.conformance_status not in CONFORMANCE_STATUSES:
            raise ArgumentError("conformance_status is invalid")
        if self.semantic_loss_status not in SEMANTIC_LOSS_STATUSES:
            raise ArgumentError("semantic_loss_status is invalid")
        if len(self.losses) > MAX_ADAPTER_EXECUTION_EVIDENCE_LOSSES or any(
            not isinstance(loss, AdapterExecutionLoss) for loss in self.losses
        ):
            raise ArgumentError("losses must contain at most 128 AdapterExecutionLoss values")
        identities = [loss.to_wire().__repr__() for loss in self.losses]
        if len(identities) != len(set(identities)):
            raise ArgumentError("loss entries must be unique")
        if self.semantic_loss_status in {"lossless", "not_applicable"} and self.losses:
            raise ArgumentError("lossless or not_applicable evidence cannot contain loss entries")
        if self.semantic_loss_status == "lossy" and not self.losses:
            raise ArgumentError("lossy evidence must contain at least one loss entry")
        if self.execution_status == "succeeded" and self.output_digest is None:
            raise ArgumentError("succeeded execution requires output_digest")
        if self.execution_status in {"refused", "failed"} and self.error_code is None:
            raise ArgumentError("refused or failed execution requires error_code")
        _bounded_optional_text("error_code", self.error_code, 128)
        if self.item_count is not None and (
            isinstance(self.item_count, bool)
            or not isinstance(self.item_count, int)
            or not 0 <= self.item_count <= MAX_ADAPTER_EXECUTION_EVIDENCE_ITEMS
        ):
            raise ArgumentError("item_count is outside its bound")
        if self.byte_length is not None and (
            isinstance(self.byte_length, bool)
            or not isinstance(self.byte_length, int)
            or not 0 <= self.byte_length <= MAX_ADAPTER_EXECUTION_EVIDENCE_BYTES
        ):
            raise ArgumentError("byte_length is outside its bound")
        if len(self.parent_digests) > MAX_ADAPTER_EXECUTION_EVIDENCE_PARENTS:
            raise ArgumentError("parent_digests must contain at most 128 values")
        for parent in self.parent_digests:
            _digest("adapter execution parent_digest", parent)
        _bounded_optional_text("attempt_id", self.attempt_id, 128)

    @classmethod
    def from_wire(cls, value: Mapping[str, Any]) -> "AdapterExecutionEvidenceRequest":
        raw = _mapping("adapter execution evidence request", value)
        allowed = {
            "group_id", "domains", "subject_id", "adapter_id", "adapter_version", "source_id", "input_digest",
            "output_digest", "execution_status", "conformance_status", "semantic_loss_status", "losses",
            "item_count", "byte_length", "error_code", "parent_digests", "attempt_id",
        }
        unknown = sorted(set(raw) - allowed)
        if unknown:
            raise ArgumentError(f"adapter execution evidence request contains unsupported fields: {', '.join(unknown)}")
        required = sorted(set(("group_id", "domains", "subject_id", "adapter_id", "adapter_version", "source_id", "input_digest", "execution_status", "conformance_status", "semantic_loss_status")) - set(raw))
        if required:
            raise ArgumentError(f"adapter execution evidence request is missing: {', '.join(required)}")
        domains = raw.get("domains")
        if not isinstance(domains, list):
            raise ArgumentError("domains must be an array")
        losses = raw.get("losses", [])
        if not isinstance(losses, list):
            raise ArgumentError("losses must be an array")
        parents = raw.get("parent_digests", [])
        if not isinstance(parents, list):
            raise ArgumentError("parent_digests must be an array")
        return cls(
            group_id=_route_text("adapter execution group_id", raw.get("group_id")),
            domains=tuple(_route_text("adapter execution domain", value) for value in domains),
            subject_id=_route_text("adapter execution subject_id", raw.get("subject_id")),
            adapter_id=_route_text("adapter execution adapter_id", raw.get("adapter_id")),
            adapter_version=_route_text("adapter execution adapter_version", raw.get("adapter_version")),
            source_id=_route_text("adapter execution source_id", raw.get("source_id")),
            input_digest=_digest("adapter execution input_digest", raw.get("input_digest")),
            output_digest=None if raw.get("output_digest") is None else _digest("adapter execution output_digest", raw.get("output_digest")),
            execution_status=_route_text("adapter execution execution_status", raw.get("execution_status")),
            conformance_status=_route_text("adapter execution conformance_status", raw.get("conformance_status")),
            semantic_loss_status=_route_text("adapter execution semantic_loss_status", raw.get("semantic_loss_status")),
            losses=tuple(AdapterExecutionLoss.from_wire(loss) for loss in losses),
            item_count=raw.get("item_count"),
            byte_length=raw.get("byte_length"),
            error_code=None if raw.get("error_code") is None else _route_text("adapter execution error_code", raw.get("error_code")),
            parent_digests=tuple(_digest("adapter execution parent_digest", value) for value in parents),
            attempt_id=None if raw.get("attempt_id") is None else _route_text("adapter execution attempt_id", raw.get("attempt_id")),
        )

    def to_mcp_arguments(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "group_id": self.group_id,
            "domains": list(self.domains),
            "subject_id": self.subject_id,
            "adapter_id": self.adapter_id,
            "adapter_version": self.adapter_version,
            "source_id": self.source_id,
            "input_digest": self.input_digest.lower(),
            "execution_status": self.execution_status,
            "conformance_status": self.conformance_status,
            "semantic_loss_status": self.semantic_loss_status,
            "losses": [loss.to_wire() for loss in self.losses],
            "parent_digests": [parent.lower() for parent in self.parent_digests],
        }
        for name in ("output_digest", "item_count", "byte_length", "error_code", "attempt_id"):
            value = getattr(self, name)
            if value is not None:
                result[name] = value.lower() if name == "output_digest" else value
        return result


@dataclass(frozen=True)
class AdapterExecutionEvidenceReport:
    raw: dict[str, Any]
    evidence: Mapping[str, Any]
    adapter: Mapping[str, Any]
    execution_status: str
    conformance_status: str
    semantic_loss_status: str
    output_digest: str | None
    evidence_digest: str
    artifact_registry: Mapping[str, Any]
    readiness_claimed: bool

    @classmethod
    def from_wire(cls, value: Mapping[str, Any]) -> "AdapterExecutionEvidenceReport":
        raw = _tool_payload(value, ADAPTER_EXECUTION_EVIDENCE_WORKFLOW)
        if raw.get("ok") is not True or raw.get("readiness_claimed") is not False:
            raise ArgumentError("adapter execution evidence is not successful or ready")
        if raw.get("execution") != "not_started" or raw.get("attestation_posture") != "caller_asserted":
            raise ArgumentError("adapter execution evidence posture is invalid")
        evidence = _mapping("adapter execution evidence", raw.get("evidence"))
        adapter = _mapping("adapter execution adapter", raw.get("adapter"))
        if evidence.get("schema") != ADAPTER_EXECUTION_EVIDENCE_SCHEMA or evidence.get("workflow") != ADAPTER_EXECUTION_EVIDENCE_WORKFLOW:
            raise ArgumentError("adapter execution evidence schema or workflow is invalid")
        execution_status = _route_text("adapter execution evidence execution_status", evidence.get("execution_status"))
        conformance_status = _route_text("adapter execution evidence conformance_status", evidence.get("conformance_status"))
        semantic_loss_status = _route_text("adapter execution evidence semantic_loss_status", evidence.get("semantic_loss_status"))
        if execution_status not in EXECUTION_STATUSES or conformance_status not in CONFORMANCE_STATUSES or semantic_loss_status not in SEMANTIC_LOSS_STATUSES:
            raise ArgumentError("adapter execution evidence contains an invalid status")
        evidence_digest = _digest("adapter execution evidence_digest", raw.get("evidence_digest"))
        if evidence.get("evidence_digest") != evidence_digest:
            raise ArgumentError("adapter execution nested and top-level evidence digests disagree")
        output_digest = evidence.get("output_digest")
        if output_digest is not None:
            output_digest = _digest("adapter execution output_digest", output_digest)
        artifact_registry = _mapping("adapter execution artifact registry", raw.get("artifact_registry"))
        if artifact_registry.get("indexed") is not True:
            raise ArgumentError("adapter execution evidence was not indexed")
        return cls(
            raw=raw,
            evidence=evidence,
            adapter=adapter,
            execution_status=execution_status,
            conformance_status=conformance_status,
            semantic_loss_status=semantic_loss_status,
            output_digest=output_digest,
            evidence_digest=evidence_digest,
            artifact_registry=artifact_registry,
            readiness_claimed=False,
        )

    def to_dict(self) -> dict[str, Any]:
        return dict(self.raw)


def adapter_execution_evidence_report(value: Mapping[str, Any]) -> AdapterExecutionEvidenceReport:
    return AdapterExecutionEvidenceReport.from_wire(value)


__all__ = [
    "ADAPTER_EXECUTION_EVIDENCE_SCHEMA",
    "ADAPTER_EXECUTION_EVIDENCE_WORKFLOW",
    "MAX_ADAPTER_EXECUTION_EVIDENCE_BYTES",
    "MAX_ADAPTER_EXECUTION_EVIDENCE_ITEMS",
    "MAX_ADAPTER_EXECUTION_EVIDENCE_LOSSES",
    "MAX_ADAPTER_EXECUTION_EVIDENCE_PARENTS",
    "AdapterExecutionLoss",
    "AdapterExecutionEvidenceRequest",
    "AdapterExecutionEvidenceReport",
    "adapter_execution_evidence_report",
]
